### PR TITLE
Optimise backend memory reads/writes

### LINKED ISSUE #1772: Optimise memory accesses on backend side
### Problem to Solve

Decrease amount of memory accesses by providing const size wrappers following existing proposal

Relates to #1771

### PR DESCRIPTION
**Pull request by breathx**
Resolves #1772 

TODO:
- [x] merge master
- [x] adapt benchmarks
- [x] adapt large_scheduled demo (rewrite in wat example)
- [x] fix ci
- [ ] re-bench (CONSIDERING)

### CODE REVIEW

File: core-backend/wasmi/src/funcs/mod.rs [hunk 1]
```diff
@@ -542,114 +474,125 @@ where
...
+                            hash: message_id.into(),
+                            ..Default::default()
+                        })
+                        .unwrap_or_else(|length| LengthWithHash {
+                            length,
+                            ..Default::default()
+                        });
+
+                    mem_ref.write_memory_as(err_mid_ptr, err_mid)
+                },
```
Comment (gshep) [inline core-backend/wasmi/src/funcs/mod.rs:489]: The code looks repetitive too. Could we have a second method, for instance `call_fallible_hash`, that executes this logic inside its body?

**Review by gshep — APPROVED** [review]
Incredible work! :fire: :gear:

File: core-backend/sandbox/src/funcs.rs [hunk 2]
```diff
@@ -122,188 +126,304 @@ where
...
-                    ctx.write_output(message_id_ptr, message_id.as_ref())
+                    ctx.write_memory_as(
+                        err_mid_ptr,
+                        LengthWithHash {
+                            hash: message_id.into(),
+                            ..Default::default()
+                        },
+                    )
+                })
+                .and_then(|length| {
```
Comment (grishasobol) [inline core-backend/sandbox/src/funcs.rs:151]: what if `error_len_on_success` returns Ok(0) ? You will overwrite `message_id` ? Also we should guaranty that error encoded size (when it is really error case) is not zero.

**Review by grishasobol — CHANGES_REQUESTED** [review]

File: core-backend/sandbox/src/funcs.rs [hunk 3]
```diff
@@ -891,14 +1267,16 @@ where
     pub fn error(ctx: &mut Runtime<E>, args: &[Value]) -> SyscallOutput {
         sys_trace!(target: "syscall::gear", "error, args = {}", args_to_str(args));
 
-        let buffer_ptr = args.iter().read()?;
+        let (error_ptr, err_ptr) = args.iter().read_2()?;
```
Comment (grishasobol) [inline core-backend/sandbox/src/funcs.rs:1270]: Comment or more clear var names  please

File: core-backend/common/src/lib.rs [hunk 4]
```diff
@@ -197,23 +197,44 @@ pub trait RuntimeCtx<E: Ext> {
...
+/// the context of calling this function, it's safe to take ptr on the
+/// object and represent it as slice. Object will be dropped after
+/// memory.write already finished execution.
+pub fn write_memory_as<T: Sized>(
+    memory: &mut impl Memory,
+    ptr: u32,
+    obj: T,
+) -> Result<(), MemoryError> {
+    let slice =
+        unsafe { slice::from_raw_parts(&obj as *const T as *const u8, mem::size_of::<T>()) };
```
Comment (ark0f) [inline core-backend/common/src/lib.rs:214]: Add comment why it's properly aligned and why is it safe to read/write types between native machine and virtual one

File: core-backend/common/src/lib.rs [hunk 5]
```diff
@@ -197,23 +197,44 @@ pub trait RuntimeCtx<E: Ext> {
...
-    fn read_memory_as<T: Sized>(&self, ptr: u32) -> Result<T, RuntimeCtxError<E::Error>> {
-        let mut buf = MaybeUninit::<T>::uninit();
-        let mut_slice =
-            unsafe { slice::from_raw_parts_mut(buf.as_mut_ptr() as *mut u8, size_of::<T>()) };
-        self.read_memory_into_buf(ptr, mut_slice)?;
-        Ok(unsafe { buf.assume_init() })
-    }
+/// Writes object in given memory as bytes.
+///
+/// Safety: due to the fact that given object is Sized and we own them in
```
Comment (ark0f) [inline core-backend/common/src/lib.rs:204]: Move safety comment above `unsafe` block. Use common format:
```
/// # Safety:
/// ...
```

**Review by ark0f — APPROVED** [review]

**Review by grishasobol — APPROVED** [review]

**Pull request merged**